File: samples/sandbox/app/src/main/java/com/scichart/docsandbox/examples/java/chartModifier2D/InteractivityTooltipsCustomization.java
package com.scichart.docsandbox.examples.java.chartModifier2D;

import android.content.Context;
import android.text.SpannableStringBuilder;
import android.util.DisplayMetrics;
import android.util.TypedValue;

import androidx.annotation.NonNull;

import com.scichart.charting.modifiers.CursorModifier;
import com.scichart.charting.modifiers.RolloverModifier;
import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting.visuals.axes.AxisInfo;
import com.scichart.charting.visuals.axes.AxisTooltip;
import com.scichart.charting.visuals.axes.DefaultAxisInfoProvider;
import com.scichart.charting.visuals.axes.IAxisTooltip;
import com.scichart.charting.visuals.axes.NumericAxis;
import com.scichart.charting.visuals.renderableSeries.FastLineRenderableSeries;
import com.scichart.charting.visuals.renderableSeries.hitTest.DefaultXySeriesInfoProvider;
import com.scichart.charting.visuals.renderableSeries.hitTest.XySeriesInfo;
import com.scichart.charting.visuals.renderableSeries.tooltips.ISeriesTooltip;
import com.scichart.charting.visuals.renderableSeries.tooltips.XySeriesTooltip;
import com.scichart.core.utility.StringUtil;
import com.scichart.docsandbox.core.ExampleDefinition;
import com.scichart.docsandbox.examples.base.SingleChart2DFragment;
import com.scichart.drawing.utility.ColorUtil;

@ExampleDefinition()
public class InteractivityTooltipsCustomization extends SingleChart2DFragment {
    @Override
    protected void initExample(@NonNull SciChartSurface surface) {}

    void useCustomSeriesInfoProvider(@NonNull SciChartSurface surface) {
        // <CustomSeriesInfoProvider>
        class CustomSeriesInfoProvider extends DefaultXySeriesInfoProvider {
            @Override
            protected ISeriesTooltip getSeriesTooltipInternal(Context context, XySeriesInfo<?> seriesInfo, Class<?> modifierType) {
                if (modifierType == CursorModifier.class) {
                    return new CustomSeriesInfoProvider.CustomXySeriesTooltip(context, seriesInfo);
                } else {
                    return super.getSeriesTooltipInternal(context, seriesInfo, modifierType);
                }
            }

            class CustomXySeriesTooltip extends XySeriesTooltip {
                public CustomXySeriesTooltip(Context context, XySeriesInfo<?> seriesInfo) {
                    super(context, seriesInfo);

                    final DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
                    final int padding = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1f, displayMetrics));
                    this.setPadding(padding, padding, padding, padding);
                }

                @Override
                protected void internalUpdate(XySeriesInfo seriesInfo) {
                    final SpannableStringBuilder sb = new SpannableStringBuilder();
                    if (seriesInfo.seriesName != null) {
                        sb.append(seriesInfo.seriesName).append(" - ");
                    }

                    sb.append("X: ").append(seriesInfo.getFormattedXValue());
                    sb.append(" Y: ").append(seriesInfo.getFormattedYValue());
                    setText(sb);

                    setTooltipBackgroundColor(0xff6495ed);
                    setTooltipStroke(0xff4d81dd);
                    setTooltipTextColor(ColorUtil.White);
                }
            }
        }
        // </CustomSeriesInfoProvider>

        // <UseCustomSeriesInfoProvider>
        final FastLineRenderableSeries line = new FastLineRenderableSeries();
        line.setSeriesInfoProvider(new CustomSeriesInfoProvider());
        // </UseCustomSeriesInfoProvider>
    }

    void useCustomAxisInfoProvider(@NonNull SciChartSurface surface) {
        // <CreateCustomAxisInfoProvider>
        class CustomAxisInfoProvider extends DefaultAxisInfoProvider {
            @Override
            protected IAxisTooltip getAxisTooltipInternal(Context context, AxisInfo axisInfo, Class<?> modifierType) {
                if (modifierType == RolloverModifier.class) {
                    return new CustomAxisTooltip(context, axisInfo);
                } else {
                    return super.getAxisTooltipInternal(context, axisInfo, modifierType);
                }
            }

            class CustomAxisTooltip extends AxisTooltip {
                public CustomAxisTooltip(Context context, AxisInfo axisInfo) {
                    super(context, axisInfo);
                    setTooltipBackground(0xff6495ed);
                }

                @Override
                protected boolean updateInternal(AxisInfo axisInfo) {
                    final SpannableStringBuilder sb = new SpannableStringBuilder();
                    sb.append("Axis ID: ").append(axisInfo.axisId).append(StringUtil.NEW_LINE);
                    sb.append("Value: ").append(axisInfo.axisFormattedDataValue);

                    setText(sb);

                    return true;
                }
            }
        }
        // </CreateCustomAxisInfoProvider>

        // <UseCustomAxisInfoProvider>
        final NumericAxis xAxis = new NumericAxis(getContext());
        xAxis.setAxisInfoProvider(new CustomAxisInfoProvider());
        // </UseCustomAxisInfoProvider>
    }
}

File: samples/sandbox/app/src/main/java/com/scichart/docsandbox/examples/kotlin/chartModifier2D/InteractivityTooltipsCustomization.kt
package com.scichart.docsandbox.examples.kotlin.chartModifier2D

import android.content.Context
import android.text.SpannableStringBuilder
import android.util.TypedValue
import com.scichart.charting.modifiers.CursorModifier
import com.scichart.charting.modifiers.RolloverModifier
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting.visuals.axes.*
import com.scichart.charting.visuals.renderableSeries.FastLineRenderableSeries
import com.scichart.charting.visuals.renderableSeries.hitTest.DefaultXySeriesInfoProvider
import com.scichart.charting.visuals.renderableSeries.hitTest.XySeriesInfo
import com.scichart.charting.visuals.renderableSeries.tooltips.ISeriesTooltip
import com.scichart.charting.visuals.renderableSeries.tooltips.XySeriesTooltip
import com.scichart.core.utility.StringUtil
import com.scichart.docsandbox.core.ExampleDefinition
import com.scichart.docsandbox.examples.base.SingleChart2DFragment
import com.scichart.drawing.utility.ColorUtil
import kotlin.math.roundToInt

@ExampleDefinition()
class InteractivityTooltipsCustomization : SingleChart2DFragment() {
    override fun initExample(surface: SciChartSurface) {}

    fun useCustomSeriesInfoProvider(surface: SciChartSurface) {
        // <CustomSeriesInfoProvider>
        class CustomSeriesInfoProvider : DefaultXySeriesInfoProvider() {
            override fun getSeriesTooltipInternal(context: Context, seriesInfo: XySeriesInfo<*>?, modifierType: Class<*>): ISeriesTooltip {
                return when (modifierType) {
                    CursorModifier::class.java -> { CustomXySeriesTooltip(context, seriesInfo) }
                    else -> { super.getSeriesTooltipInternal(context, seriesInfo, modifierType) }
                }
            }

            inner class CustomXySeriesTooltip(context: Context?, seriesInfo: XySeriesInfo<*>?) : XySeriesTooltip(context, seriesInfo) {
                init {
                    val displayMetrics = resources.displayMetrics
                    val padding = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1f, displayMetrics).roundToInt()
                    setPadding(padding, padding, padding, padding)
                }

                override fun internalUpdate(seriesInfo: XySeriesInfo<*>) {
                    val sb = SpannableStringBuilder()
                    seriesInfo.seriesName?.run {
                        sb.append(this).append(StringUtil.NEW_LINE)
                    }
                    sb.append("X: ").append(seriesInfo.formattedXValue)
                    sb.append(" Y: ").append(seriesInfo.formattedYValue)
                    text = sb

                    setTooltipBackgroundColor(0xff6495ed.toInt())
                    setTooltipStroke(0xff4d81dd.toInt())
                    setTooltipTextColor(ColorUtil.White)
                }
            }
        }
        // </CustomSeriesInfoProvider>

        // <UseCustomSeriesInfoProvider>
        val line = FastLineRenderableSeries()
        line.seriesInfoProvider = CustomSeriesInfoProvider()
        // </UseCustomSeriesInfoProvider>
    }

    fun useCustomAxisInfoProvider(surface: SciChartSurface) {
        // <CreateCustomAxisInfoProvider>
        class CustomAxisInfoProvider : DefaultAxisInfoProvider() {
            override fun getAxisTooltipInternal(context: Context, axisInfo: AxisInfo, modifierType: Class<*>): IAxisTooltip {
                return when (modifierType) {
                    RolloverModifier::class.java -> { CustomAxisTooltip(context, axisInfo) }
                    else -> { super.getAxisTooltipInternal(context, axisInfo, modifierType) }
                }
            }

            inner class CustomAxisTooltip(context: Context?, axisInfo: AxisInfo?) : AxisTooltip(context, axisInfo) {
                override fun updateInternal(axisInfo: AxisInfo): Boolean {
                    val sb = SpannableStringBuilder()
                    sb.append("Axis ID: ").append(axisInfo.axisId).append(StringUtil.NEW_LINE)
                    sb.append("Value: ").append(axisInfo.axisFormattedDataValue)

                    text = sb

                    return true
                }

                init {
                    setTooltipBackground(0xff6495ed.toInt())
                }
            }
        }
        // </CreateCustomAxisInfoProvider>

        // <UseCustomAxisInfoProvider>
        val xAxis = NumericAxis(context)
        xAxis.axisInfoProvider = CustomAxisInfoProvider()
        // </UseCustomAxisInfoProvider>
    }
}

